src/ucp/core/ucp_request.c

```diff
@@ -21,6 +21,81 @@
 
 const ucp_request_param_t ucp_request_null_param = { .op_attr_mask = 0 };
 
+static inline uint8_t ucp_request_get_mem_type(ucp_request_t *req)
+{
+    if (req->flags & (UCP_REQUEST_FLAG_SEND_AM | UCP_REQUEST_FLAG_SEND_TAG)) {
+        if (req->flags & UCP_REQUEST_FLAG_PROTO_INITIALIZED) {
+            return req->send.state.dt_iter.mem_info.type;
+        } else {
+            return req->send.mem_type;
+        }
+    } else if (req->flags &
+               (UCP_REQUEST_FLAG_RECV_AM | UCP_REQUEST_FLAG_RECV_TAG)) {
+        return req->recv.mem_type;
+    } else {
+        return UCS_MEMORY_TYPE_UNKNOWN;
+    }
+}
+
+static void ucp_request_print(ucs_string_buffer_t *strb, ucp_request_t *req)
+{
+    ucp_ep_h ep;
+    ucp_ep_config_t *config;
+
+    if (req->flags & (UCP_REQUEST_FLAG_SEND_AM | UCP_REQUEST_FLAG_SEND_TAG)) {
+        ucs_string_buffer_appendf(
+                strb, "send mem_type: %s len: %zu ",
+                ucs_memory_type_names[ucp_request_get_mem_type(req)],
+                req->send.length);
+
+        if (req->flags & UCP_REQUEST_FLAG_PROTO_INITIALIZED) {
+            ucp_proto_select_param_str(&req->send.proto_config->select_param,
+                                       strb);
+        }
+
+        ep     = req->send.ep;
+        config = ucp_ep_config(ep);
+        ucp_ep_config_lane_info_str(ep->worker, &config->key, NULL,
+                                    req->send.lane, UCP_NULL_RESOURCE, strb);
+    } else if (req->flags &
+               (UCP_REQUEST_FLAG_RECV_AM | UCP_REQUEST_FLAG_RECV_TAG)) {
+        ucs_string_buffer_appendf(
```

Comment:
1. "mem_type %s" -> "%s memory" (e.g "host memory", "cuda memory", etc)
2. can move memory type part outside if/else
3. maybe format like this: "send length 4194304, host memory"